redash/handlers/query_results.py

```diff
@@ -98,6 +99,11 @@ def get(self, query_id=None, query_result_id=None, filetype='json'):
         if query_result_id:
             query_result = models.QueryResult.get_by_id(query_result_id)
 
+        if not query:
+            query = models.Query.get(query_hash=query_result.query_hash)
+        if not (set(query.access_groups) & set(current_user.groups)):
```

Comment:
This prevents from running adhoc (non saved) queries... 
I suggest the following logic:
1. If there is query specified, check if the user has access to it.
2. If there is no query specified, check if the user has access to the data source.